airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/DestinationTaskLauncher.kt

```diff
@@ -81,13 +87,53 @@ class DefaultDestinationTaskLauncher(
     private val processRecordsTaskFactory: ProcessRecordsTaskFactory,
     private val processBatchTaskFactory: ProcessBatchTaskFactory,
     private val closeStreamTaskFactory: CloseStreamTaskFactory,
-    private val teardownTaskFactory: TeardownTaskFactory
+    private val teardownTaskFactory: TeardownTaskFactory,
+    private val exceptionHandler: DestinationTaskLauncherExceptionHandler
 ) : DestinationTaskLauncher {
     private val log = KotlinLogging.logger {}
 
     private val runTeardownOnce = AtomicBoolean(false)
     private val batchUpdateLock = Mutex()
 
+    class SyncTaskWrapper(
+        private val exceptionHandler: DestinationTaskLauncherExceptionHandler,
+        private val innerTask: Task,
+    ) : Task {
+        override suspend fun execute() {
+            try {
+                innerTask.execute()
+            } catch (t: Throwable) {
+                exceptionHandler.handleSyncFailure(t)
+            }
+        }
+    }
+
+    class StreamTaskWrapper(
+        private val exceptionHandler: DestinationTaskLauncherExceptionHandler,
+        private val innerTask: Task,
+    ) : Task {
+        override suspend fun execute() {
+            try {
+                innerTask.execute()
+            } catch (t: Throwable) {
+                exceptionHandler.handleStreamFailure(t)
+            }
+        }
+    }
+
+    private suspend fun enqueue(task: DestinationWriteTask) {
+        taskRunner.enqueue(
+            when (task) {
```

Comment:
can we make `Task` a sealed interface? and/or does this want an `else` clause?

Reply:
Hm. Yeah, I've been trying to decide whether other things should launch tasks (like checkpoint flushing), so it would make sense to have like

```
Task -> DestinationWriteTask -> SyncTask, StreamTask
     \-> SomeotherName -> FlushCheckpointTask, etc
```